internal/types/clone_url.go

```
@@ -0,0 +1,277 @@
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/awscodecommit"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/bitbucketcloud"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/bitbucketserver"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/github"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/gitlab"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/gitolite"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/phabricator"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+// RepoCloneURL builds a cloneURl for the given repo based on the
+// external service configuration.
+func RepoCloneURL(kind, config string, repo *Repo) (string, error) {
+	parsed, err := extsvc.ParseConfig(kind, config)
+	if err != nil {
+		return "", errors.Wrap(err, "loading service configuration")
+	}
+
+	switch t := parsed.(type) {
+	case *schema.AWSCodeCommitConnection:
+		return awsCodeCloneURL(repo, t), nil
+	case *schema.BitbucketServerConnection:
+		return bitbucketServerCloneURL(repo, t), nil
+	case *schema.BitbucketCloudConnection:
+		return bitbucketCloudCloneURL(repo, t), nil
+	case *schema.GitHubConnection:
+		return githubCloneURL(repo, t)
+	case *schema.GitLabConnection:
+		return gitlabCloneURL(repo, t), nil
+	case *schema.GitoliteConnection:
+		return gitoliteCloneURL(repo, t), nil
+	case *schema.PerforceConnection:
+		return perforceCloneURL(repo, t), nil
+	case *schema.PhabricatorConnection:
+		return phabricatorCloneURL(repo, t), nil
+	case *schema.OtherExternalServiceConnection:
+		return otherCloneURL(repo, t), nil
+	default:
+		return "", fmt.Errorf("unknown external service kind %q", kind)
+	}
+}
+
+func awsCodeCloneURL(repo *Repo, cfg *schema.AWSCodeCommitConnection) string {
+	metadata := repo.Metadata.(*awscodecommit.Repository)
+	u, err := url.Parse(metadata.HTTPCloneURL)
+	if err != nil {
+		log15.Warn("Error adding authentication to AWS CodeCommit repository Git remote URL.", "url", metadata.HTTPCloneURL, "error", err)
```

I find it very unusual to print a log in a utility function... same in other places within the file.

---

This is to mimic the behavior from the repo-updater Sources code which skips errors and tries to generate a clone url in all cases. I also find it unusual and would rather return an error and let the caller deal with it, but I didn't want to make a change to big.
What do you think? I'm open to any suggestion

---

I agree that we should mimic changes, but we should have an issue filed for tracking it or a direct PR followup :)